pkg/file/joiner/internal/job.go

```diff
@@ -0,0 +1,211 @@
+// nextReference gets the next chunk reference at the cursor of the chunk currently loaded
+// for the specified level.
+func (j *SimpleJoinerJob) nextReference(level int) error {
+	data := j.data[level]
+	cursor := j.cursors[level]
+	chunkAddress := swarm.NewAddress(data[cursor : cursor+swarm.SectionSize])
+	err := j.nextChunk(level-1, chunkAddress)
+	if err != nil {
+		return err
+	}
+
+	// move the cursor to the next reference
+	j.cursors[level] += swarm.SectionSize
+	return nil
+}
+
+// nextChunk retrieves data chunks by resolving references in intermediate chunks.
+// The method will be called recursively via the nextReference method when
+// the current chunk is an intermediate chunk.
+// When a data chunk is found it is passed on the dataC channel to be consumed by the
+// io.Reader consumer.
+func (j *SimpleJoinerJob) nextChunk(level int, address swarm.Address) error {
+
+	// attempt to retrieve the chunk
+	j.logger.Tracef("next chunk level %d get: %v", level, address)
+	ch, err := j.store.Get(j.ctx, storage.ModeGetRequest, address)
+	if err != nil {
+		return err
+	}
+	j.cursors[level] = 0
+	j.data[level] = ch.Data()[8:]
+
+	// any level higher than 0 means the chunk contains references
+	// which must be recursively processed
+	if level > 0 {
+		j.logger.Tracef("cursor %d datalen %d", j.cursors[level], len(j.data[level]))
+		for j.cursors[level] < len(j.data[level]) {
+			if len(j.data[level]) == j.cursors[level] {
+				j.data[level] = ch.Data()[8:]
+				j.cursors[level] = 0
+			}
+			err = j.nextReference(level)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		// read data and pass to reader only if session is still active
+		// * context cancelled when client has disappeared, timeout etc
+		// * doneC receive when gracefully terminated through Close
+		data := ch.Data()[8:]
+		select {
+		case <-j.ctx.Done():
+			j.logger.Tracef("context done %v", j.ctx.Err())
+			j.readCount = j.spanLength
+			return j.ctx.Err()
+		case <-j.doneC:
+			return file.NewErrAbort(errors.New("chunk read aborted"))
+		case j.dataC <- data:
+			j.readCount += int64(len(data))
+		}
+		// when we reach the end of data to be read
+		// bubble io.EOF error to the gofunc in the 
+		// constructor that called start()
+		if j.readCount == j.spanLength {
+			j.logger.Trace("read all")
+			return io.EOF
+		}
+	}
+	return err
+}
+
+// Read is called by the consumer to retrieve the joined data.
+// It must be called with a buffer equal to the maximum chunk size.
+func (j *SimpleJoinerJob) Read(b []byte) (n int, err error) {
+	if cap(b) != swarm.ChunkSize {
+		return 0, fmt.Errorf("Read must be called with a buffer of %d bytes", swarm.ChunkSize)
+	}
+	data, ok := <-j.dataC
+	if !ok {
+		<-j.doneC
+		return 0, j.err
+	}
+	j.logger.Tracef("Read %d %x", len(data), data[:16])
+	copy(b, data)
+	return len(data), nil
+}
+
+// Close is called by the consumer to gracefully abort the data retrieval.
+func (j *SimpleJoinerJob) Close() error {
+	j.doneC <- struct{}{}
```

Comment:
Should this be close(j.doneC)? Sending to this channel will unblock only one goroutine which reads it. But this is a problem because this channel is closed also in the goroutine created in the constructor. So there are two gorutines that own this chunk. In this particular case, I would suggest using sync.Once to make sure that the channel is closed only once.